test/evaluate/intrinsics.cc

```diff
@@ -0,0 +1,214 @@
+    keywords.push_back("");
+    return *this;
+  }
+  template<typename A> TestCall &Push(NamedArg<A> &&x) {
+    args.emplace_back(AsGenericExpr(std::move(x.value)));
+    keywords.push_back(x.keyword);
+    strings.Save(x.keyword);
+    return *this;
+  }
+  template<typename A, typename... As> TestCall &Push(A &&x, As &&... xs) {
+    Push(std::move(x));
+    return Push(std::move(xs)...);
+  }
+  void Marshal() {
+    strings.Save(name);
+    strings.Marshal();
+    std::size_t j{0};
+    for (auto &kw : keywords) {
+      if (!kw.empty()) {
+        args[j].keyword = strings(kw);
+      }
+      ++j;
+    }
+  }
+  void DoCall(std::optional<DynamicType> resultType = std::nullopt,
+      int rank = 0, bool isElemental = false) {
+    Marshal();
+    parser::CharBlock fName{strings(name)};
+    std::cout << "function: " << fName.ToString();
+    char sep{'('};
+    for (const auto &a : args) {
+      std::cout << sep;
+      sep = ',';
+      a.Dump(std::cout);
+    }
+    if (sep == '(') {
+      std::cout << '(';
+    }
+    std::cout << ")\n";
+    CallCharacteristics call{fName, args};
+    auto messages{strings.Messages(buffer)};
+    std::optional<SpecificIntrinsic> si{table.Probe(call, &messages)};
+    if (resultType.has_value()) {
+      TEST(si.has_value());
+      TEST(buffer.empty());
+      TEST(*resultType == si->type);
+      MATCH(rank, si->rank);
+      MATCH(isElemental, si->isElemental);
+    } else {
+      TEST(!si.has_value());
+      TEST(!buffer.empty() || name == "bad");
+    }
+    strings.Emit(std::cout, buffer);
+  }
+
+  const IntrinsicProcTable &table;
+  CookedStrings strings;
+  parser::Messages buffer;
+  Arguments args;
+  std::string name;
+  std::vector<std::string> keywords;
+};
+
+template<typename A> void Push(Arguments &args, A &&x) {
+  args.emplace_back(AsGenericExpr(std::move(x)));
+}
+template<typename A, typename... As>
+void Push(Arguments &args, A &&x, As &&... xs) {
+  args.emplace_back(AsGenericExpr(std::move(x)));
+  Push(args, std::move(xs)...);
+}
+template<typename... As> Arguments Args(As &&... xs) {
+  Arguments args;
+  Push(args, std::move(xs)...);
+  return args;
+}
+
+void TestIntrinsics() {
+  IntrinsicTypeDefaultKinds defaults;
+  MATCH(4, defaults.defaultIntegerKind);
+  MATCH(4, defaults.defaultRealKind);
+  IntrinsicProcTable table{IntrinsicProcTable::Configure(defaults)};
+  table.Dump(std::cout);
+
+  using Int1 = Type<TypeCategory::Integer, 1>;
+  using Int4 = Type<TypeCategory::Integer, 4>;
+  using Int8 = Type<TypeCategory::Integer, 8>;
+  using Real4 = Type<TypeCategory::Real, 4>;
+  using Real8 = Type<TypeCategory::Real, 8>;
+  using Complex4 = Type<TypeCategory::Complex, 4>;
+  using Complex8 = Type<TypeCategory::Complex, 8>;
+  using Char = Type<TypeCategory::Character, 1>;
+  using Log4 = Type<TypeCategory::Logical, 4>;
+
+  TestCall{table, "bad"}
+      .Push(Const(Scalar<Int4>{}))
+      .DoCall();  // bad intrinsic name
```

Comment:
How are the results verified? E.g. in this case that "bad" is not an intrinsic?

Reply:
There's no valid return type specified as an argument to `DoCall`, so intrinsic table probing must fail and must generate a message.